File: e-java/src/main/java/e/java/JsonStringEncoder.java
package e.java;

import java.util.Map;
import java.util.StringJoiner;

public final class JsonStringEncoder implements Encoder<String> {
    private static JsonStringEncoder instance;

    private JsonStringEncoder() {}

    public static JsonStringEncoder get() {
        if (instance == null) {
            instance = new JsonStringEncoder();
        }

        return instance;
    }

    @Override public String encode(E e) {
        StringJoiner joiner = new StringJoiner(",", "{", "}");

        if (e.hasName())    joiner.add(String.format("\"name\":\"%s\"",    escape(e.name())));
        if (e.hasMessage()) joiner.add(String.format("\"message\":\"%s\"", escape(e.message())));
        if (e.hasCode())    joiner.add(String.format("\"code\":%d",        e.code()));;
        if (e.hasCause())   joiner.add(String.format("\"cause\":%s",       encodeCause(e)));
        if (e.hasData())    joiner.add(String.format("\"data\":%s",        encodeData(e)));

        return joiner.toString();
    }

    private String encodeData(E e) {
        StringJoiner joiner = new StringJoiner(",", "{", "}");

        for (Map.Entry<String, String> entry : e.data().entrySet()) {
            joiner.add(String.format("\"%s\":\"%s\"", escape(entry.getKey()), escape(entry.getValue())));
        }

        return joiner.toString();
    }

    private String encodeCause(E e) {
        return e.cause() == null ? "null" : String.format("\"%s\"", escape(e.cause().getMessage()));
    }

    private String escape(String s) {
        return s.replace("\"", "\\\"");
    }
}

File: e-kotlin/src/main/kotlin/e/kotlin/JsonStringEncoder.kt
package e.kotlin

import java.util.*

object JsonStringEncoder : Encoder<String> {
    override fun encode(e: E): String {
        val joiner = StringJoiner(",", "{", "}")

        if (e.hasName())    joiner.add(""""name":"${escape(e.name())}"""")
        if (e.hasMessage()) joiner.add(""""message":"${escape(e.message())}"""")
        if (e.hasCode())    joiner.add(""""code":${e.code()}""")
        if (e.hasCause())   joiner.add(""""cause":${encodeCause(e.cause())}""")
        if (e.hasData())    joiner.add(""""data":${encodeData(e.data())}""")

        return joiner.toString()
    }

    private fun encodeCause(cause: Throwable?): String =
        if (cause?.message == null) "null" else """"${escape(cause.message!!)}""""

    private fun encodeData(data: Map<String, String>): String =
        data.map { e -> """"${escape(e.key)}":"${escape(e.value)}"""" }.joinToString(",", "{", "}")

    private fun escape(s: String): String = s.replace("\"", "\\\"")
}
